File: Day #168/solution.js
// Rotate NxN 90 clockwise in place: transpose then reverse each row. O(n^2) time, O(1) extra space.

function rotate(m) {
    const n = m.length;
    for (let i = 0; i < n; i++)
        for (let j = i + 1; j < n; j++)
            [m[i][j], m[j][i]] = [m[j][i], m[i][j]];
    for (let i = 0; i < n; i++) m[i].reverse();
}

function main() {
    const m = [[1, 2, 3], [4, 5, 6], [7, 8, 9]];
    rotate(m);
    for (const row of m) console.log("[" + row.join(", ") + "]");
}

main();

File: Day #168/solution.go
// Rotate NxN 90 clockwise in place: transpose then reverse each row. O(n^2) time, O(1) extra space.
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func rotate(m [][]int) {
	n := len(m)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			m[i][j], m[j][i] = m[j][i], m[i][j]
		}
	}
	for i := 0; i < n; i++ {
		for a, b := 0, n-1; a < b; a, b = a+1, b-1 {
			m[i][a], m[i][b] = m[i][b], m[i][a]
		}
	}
}

func main() {
	m := [][]int{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}}
	rotate(m)
	for _, row := range m {
		parts := make([]string, len(row))
		for j, v := range row {
			parts[j] = strconv.Itoa(v)
		}
		fmt.Println("[" + strings.Join(parts, ", ") + "]")
	}
}
